### ISSUE TITLE
Can we get more equivalencies from the KG2 build process?

### ISSUE THREAD

**Issue by edeutsch** [RTXteam/RTX#888]
Now that we have a nifty NodeSynonymizer that has quite a lot of information, more than the SRI normalizer I think, I wonder how we could make it better. It currently draws its associations from two places:
1) SRI node normalizer. I don't know where they get their data
2) From KG2 for nodes that have exactly* the same name

I think the gene/protein equivalencies could be expanded and we could get that from UniProtKB.

But I wonder if as a side product of ETL to KG2, there are synonyms that could perhaps be dumped to a file of synonyms? I suppose that KG2 ETL is most concerned with relationships, but probably synonym relationships are mostly ignored? but maybe they could be dumped to a file that the NodeNormalizer could read to potentially expand its world of synonyms, especially in cases where names are not identical?

What do you think?

**Comment by dkoslicki** [comment]
@edeutsch I'd ping @saramsey to see about this issue about getting more equivalencies

**Comment by saramsey** [comment]
The new KG2 will indicate many synonyms using the `biolink:same_as` relationship CURIE. I suppose that in addition to (1) and (2) above, we could enumerate the `same_as` edges to get a bunch more synonym relationships.

Further, KG2 has a `synonym` list-type node property. We could parse CURIE IDs out of that as well. Warning: the `synonym` list is kind of wild-west. But a simple filter (using the allowed prefixes in `RTX/code/kg2/curies-to-urls-map.yaml` in the `kg2-curie-refactoring` branch) should be able to identify valid CURIE IDs from the synonym list.

**Comment by edeutsch** [comment]
Well, if someone would build me a file that contains any synonyms that emerge as a byproduct of KG2 building, I could incorporate it into the NodeSynonymizer to make it smarter.

**Comment by edeutsch** [comment]
and discussed today, if someone is willing to create an export script (like dumpdata.py) that dumps to a file all equivalence relationships in KG2, then the NodeSynonymizer can slurp that in and improve its equivalence mapping.

**Comment by amykglen** [comment]
I just added a simple script [dump_kg2_equivalencies.py](https://github.com/RTXteam/RTX/blob/master/code/ARAX/NodeSynonymizer/dump_kg2_equivalencies.py) that does this, @edeutsch (grabs node pairs connected by an `equivalent_to` edge in KG2). I stuck it in the `NodeSynonymizer` directory, but feel free to move it. It currently creates a TSV file with the results (~350k rows), but output format could easily be changed. It doesn't take long to run - 15 seconds or so.

(Note: Once the new KG2 build is rolled out, I think the script will need to be updated to look for `same_as` relationships rather than `equivalent_to` relationships. But that'll be a very minor adjustment.)

**Comment by amykglen** [comment]
Updated the script so that it will work with the new KG2 build (it's backwards-compatible though).

**Comment by edeutsch** [comment]
thanks, I will update the NodeSynonymizer to use this!

**Comment by saramsey** [comment]
We have 350k `same_as` relationships in the latest KG2 build

<img width="337" alt="Screen Shot 2020-07-16 at 11 16 11 AM" src="https://user-images.githubusercontent.com/5562107/87707340-cd602800-c755-11ea-8602-ed95f9858597.png">

**Comment by amykglen** [comment]
nice - fyi, @saramsey - I also see some `equivalent_to` relationships in there still. not sure if that's intended. the script I wrote for @edeutsch considers both though.

![Screen Shot 2020-07-16 at 11 38 17 AM](https://user-images.githubusercontent.com/49423686/87709281-e0c0c280-c758-11ea-8391-ba20d7d27801.png)

**Comment by amykglen** [comment]
@edeutsch - I also just added a second script ([dump_kg2_synonym_field.py](https://github.com/RTXteam/RTX/blob/master/code/ARAX/NodeSynonymizer/dump_kg2_synonym_field.py)) that extracts the contents of the `synonym` property on KG2 nodes, and puts the results in a JSON file. so the output JSON looks like this:
```
{
   "CUI:C0009392":[
      "Myeloid Cell-Growth Inducer",
      "MGI-1",
      "Colony-Stimulating Factors",
      "Colony Stimulating Factor",
      "Colony-Stimulating Factor",
      "MGI-1 Protein"
   ],
   "ATC:C08CA":[
      "Dihydropyridine derivatives"
   ],
   "CUI:C0301556":[
      "arginine hydrochloride"
   ],
   "CUI:C0024337":[
      "L-Lysine",
      "lysine",
      "LYSINE",
      "Lysine"
   ],
   "CUI:C0035923":[
      "RUBELLA VIRUS VACCINE,LIVE",
      "Rubella Vaccine"
   ],
   "CUI:C0025029":[
      "Mecamylamine",
      "Bicyclo(2.2.1)heptan-2-amine, N,2,3,3-tetramethyl-",
      "MECAMYLAMINE"
   ]
}
```
but includes about 4 million nodes (all nodes in KG2 with a non-empty `synonym` property). seems like useful info for the `NodeSynonymizer`..

it takes a few minutes to run. output file is about 900MB.

**Comment by saramsey** [comment]
> nice - fyi, @saramsey - I also see some `equivalent_to` relationships in there still. not sure if that's intended. the script I wrote for @edeutsch considers both though.
> 
> ![Screen Shot 2020-07-16 at 11 45 37 AM](https://user-images.githubusercontent.com/49423686/87709946-e66ad800-c759-11ea-814f-0a8e4993393c.png)

Thank you for bringing this to my attention. I have noted this as #963. Already working on a fix.

**Comment by saramsey** [comment]
> > nice - fyi, @saramsey - I also see some `equivalent_to` relationships in there still. not sure if that's intended. the script I wrote for @edeutsch considers both though.
> > ![Screen Shot 2020-07-16 at 11 45 37 AM](https://user-images.githubusercontent.com/49423686/87709946-e66ad800-c759-11ea-814f-0a8e4993393c.png)
> 
> @amykglen Thank you for bringing this to my attention. I have noted this as #963. Already working on a fix.

Tentative fix committed (see #963).

**Comment by saramsey** [comment]
Here is where we are at, with the latest KG2 build (on `kg2endpoint.rtx.ai`):
<img width="650" alt="Screen Shot 2020-07-23 at 10 41 34 AM" src="https://user-images.githubusercontent.com/5562107/88319584-38b77600-ccd1-11ea-8d35-6a66831d36dd.png">

**Comment by saramsey** [comment]
OK, can I close this issue out?

**Comment by dkoslicki** [comment]
> OK, can I close this issue out?

@edeutsch and @amykglen as they want/need more equivalences and may not have seen this comment by Steve

**Comment by edeutsch** [comment]
My status is that I still need to fully integrate the equivalencies that we are already able to dump via Amy's exporters. I've made good progress, but am having a problem with the merging that I need to debug.

Beyond that, I suppose the question is: have we wrung the KG2 build process dry? No more equivalencies to be had beyond what is already included?
If so, this can be closed, yes.

**Comment by amykglen** [comment]
as far as I know, we've extracted all such info from KG2. (`same_as`/`equivalent_to` relationships and the contents of the `synonym` property.) I haven't thought of other relevant info to extract.

**Comment by edeutsch** [comment]
Is it a standing policy that all equivalency information is ETLed from all information sources (or if not *all* then like 90%)? Or does that just occur for a few?

**Comment by saramsey** [comment]
> Is it a standing policy that all equivalency information is ETLed from all information sources (or if not _all_ then like 90%)? Or does that just occur for a few?

Are you asking if we could get more `biolink:same_as` relations by transitive closure?  I would think we could get tons via transitive closure, right @amykglen ?

**Comment by amykglen** [comment]
true! that's something I could look into... although, perhaps the way @edeutsch incorporates the info in the `same_as` dump into the synonymizer already achieves that? (since I'm guessing that as each `same_as` node pair is encountered, their synonym groups are joined? thus as you move down the list, `p` is eventually added to `n`'s synonym group if `(n)-[:same_as]-(m)-[:same_as]-(p)`, and etc.)

I'm wondering if @edeutsch was more asking about whether when `same_as` relationship info is available in the upstream knowledge sources KG2 ingests (e.g., Drugbank, NCBIGene, etc..), if we always extract it and incorporate it into KG2?

**Comment by edeutsch** [comment]
> true! that's something I could look into... although, perhaps the way @edeutsch incorporates the info in the same_as dump into the synonymizer already achieves that? (since I'm guessing that as each same_as node pair is encountered, their synonym groups are joined? thus as you move down the list, p is eventually added to n's synonym group if (n)-[:same_as]-(m)-[:same_as]-(p), and etc.)

It is supposed to work this way, but it is currently not and I don't understand why. I need more debugging time..

> I'm wondering if @edeutsch was more asking about whether when same_as relationship info or synonym info is available in the upstream knowledge sources KG2 ingests (e.g., Drugbank, NCBIGene, etc..), if we always extract it and incorporate it into KG2?

yes, this was my question.

**Comment by saramsey** [comment]
FWIW, there are 558,524 `same_as` concept equivalencies in the new KG2.3.0:

<img width="667" alt="Screen Shot 2020-08-20 at 9 06 57 PM" src="https://user-images.githubusercontent.com/5562107/90851318-2b6dc580-e329-11ea-9923-7d50a0ebf994.png">

**Comment by finnagin** [comment]
@amykglen @edeutsch is this still relevant or can we close?
